gateway-service/src/main/java/org/zowe/apiml/gateway/config/AdditionalRegistrationConfig.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ */
+
+package org.zowe.apiml.gateway.config;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.env.StandardEnvironment;
+import org.springframework.util.CollectionUtils;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+@Slf4j
+@Configuration
+@RequiredArgsConstructor
+public class AdditionalRegistrationConfig {
+
+    public static final String DISCOVERY_SERVICE_URLS_KEY = "DISCOVERYSERVICEURLS";
+    public static final String GATEWAY_URL_KEY = "GATEWAYURL";
+    public static final String COMMON_PREFIX = "ZWE_CONFIGS_APIML_SERVICE_ADDITIONALREGISTRATION_";
+    public static final String SERVICE_URL_KEY = "SERVICEURL";
+    private static final int EXPECTED_ROUTE_PART_INDEX = 2;
+
+    @Bean
+    public List<AdditionalRegistration> additionalRegistration(StandardEnvironment environment) {
+        List<AdditionalRegistration> additionalRegistrations = extractAdditionalRegistrations(System.getenv());
+        log.debug("Parsed {} additional regs, \t first: {}", additionalRegistrations.size(), additionalRegistrations.stream().findFirst().orElse(null));
+        return additionalRegistrations;
+    }
+
+    static List<AdditionalRegistration> extractAdditionalRegistrations(Map<String, String> allProperties) {
+
+        if (CollectionUtils.isEmpty(allProperties)) {
+            return Collections.emptyList();
+        }
+
+        Map<String, String> additionalProperties = allProperties.entrySet().stream().filter(entry -> entry.getKey().startsWith(COMMON_PREFIX))
```

Comment:
Do you think it is the best way to parse it? It is too complex and I guess it is not necessary.

Basically, there are two ways to achieve it:

1. iterate till a value exists:

List<AdditionalRegistration> ars = new LinkedList<>();
for (int i = 0; ;i++) {
 boolean found = false;
 AdditionalRegistration ar = new AdditionalRegistration();
 
 String dsList = System.getProperty(String.format("%s[%s]_DISCOVERYSERVICEURLS", COMMON_PREFIX, i);
 if (ar |= !StringUtils.isEmpty(dsList)) {
  ar.setDiscoveryServiceUrls(dsList);
 }

 if (found) {
  ars.add(ar);
 } else {
  break;
 }

}

2. Iterate over system properties

private static final PATTERN_DS_URLS = Pattern.compile("^ZWE_CONFIGS_APIML_SERVICE_ADDITIONALREGISTRATION_\\[([0-9]+)\\]_DISCOVERYSERVICEURLS");

private AdditionalRegistration getAdditionalRegistration (Map<Integer, AdditionalRegistration> ars, String key, Pattern pattern) {
 Matcher matcher = patcher.matches(key);
 if (matcher.matches()) {
  int index = Integer.parseInt(matcher.group(1));
  return ars.computeIfAbsent(index, AdditionalRegistration::new);
 }
 return null;
}

private boolean processDsUrl(String key, Map<Integer, AdditionalRegistration> ars)  {
 AdditionalRegistration ar = getAdditionalRegistration (key, PATTERN_DS_URLS);
 if (ar != null) {
  ar.setDiscoveryServiceUrls(System.getProperty(key);
  return true;
 }
 return false;
}

Map<Integer, AdditionalRegistration> ars = new HashMap<>();

System.getProperties().keySet().stream().map(String.class::cast).filter(s -> s.startsWith()).forEach(key -> 
  processDsUrl(key, ars) ||
  process....
);

I guess both solutions are shorter, easier to read, and definitely safer for NullPointerException, IndexOutOfBound, etc.

Reply:
Ok, I've simplified it with regexes.